server/test/lib/scheduler/UserSchedule.spec.ts

```diff
@@ -1,86 +1,217 @@
+        await this.userSchedule.assignWorkspace(workspaces.get("2"));
+        const mostRecentAssignment = this.userSchedule.getMostRecentAssignment()
+        expect(mostRecentAssignment.getWorkspace()).to.equal(workspaces.get("2"));
       });
     });
   });
 
-  describe("getMostRecentAssignment method", function() {
+  describe("getTreesWorkedOnLeastRecentlyByUser method", function() {
+    const rootWorkspaces = [
+      workspaces.get("1"),
+      workspaces.get("2"),
+      workspaces.get("3"),
+      workspaces.get("4"),
+      workspaces.get("5"),
+    ];
+
     context("with a user that has not been assigned to workspace", function() {
-      it("returns undefined", function() {
-        expect(this.userSchedule.getMostRecentAssignment()).to.equal(undefined);
+      it("returns all trees", function() {
+        const trees = this.userSchedule.getTreesWorkedOnLeastRecentlyByUser(rootWorkspaces);
+        expect(trees).to.have.deep.members(rootWorkspaces);
       });
     });
 
-    context("with a user that has been assigned a workspace once", function() {
-      it("returns that workspace", function() {
-        this.userSchedule.assignWorkspace(workspaces[0]);
-        const mostRecentAssignment = this.userSchedule.getMostRecentAssignment()
-        expect(mostRecentAssignment.getWorkspace()).to.equal(workspaces[0]);
+    context("with a user that has been assigned to some trees", function() {
+      it("returns correct trees", async function() {
+        await this.userSchedule.assignWorkspace(workspaces.get("2"));
+        const trees = this.userSchedule.getTreesWorkedOnLeastRecentlyByUser(rootWorkspaces);
+        expect(trees).to.have.deep.members(rootWorkspaces.filter(w => w.id !== "2"));
       });
     });
 
-    context("with a user that has been assigned two workspaces", function() {
-      it("returns the most recently assigned workspace", function() {
-        this.userSchedule.assignWorkspace(workspaces[1]);
-        this.userSchedule.assignWorkspace(workspaces[2]);
-        const mostRecentAssignment = this.userSchedule.getMostRecentAssignment()
-        expect(mostRecentAssignment.getWorkspace()).to.equal(workspaces[2]);
+    context("with a user that has been assigned to all trees", function() {
+      it("returns correct tree", async function() {
+        await this.userSchedule.assignWorkspace(workspaces.get("1"));
+        this.clock.tick(ONE_MINUTE);
+        await this.userSchedule.assignWorkspace(workspaces.get("2-1"));
+        this.clock.tick(ONE_MINUTE);
+        await this.userSchedule.assignWorkspace(workspaces.get("3"));
+        this.clock.tick(ONE_MINUTE);
+        await this.userSchedule.assignWorkspace(workspaces.get("1-1-1"));
+        this.clock.tick(ONE_MINUTE);
+        await this.userSchedule.assignWorkspace(workspaces.get("5-2"));
+        this.clock.tick(ONE_MINUTE);
+        await this.userSchedule.assignWorkspace(workspaces.get("4"));
```

Comment:
this code really could be dryed up a lot. For instance, two of these lines could look like,
```
increment_time(ONE_MINUTE)
await assignWorkspace("4")
```